connection.go

```diff
@@ -62,15 +64,55 @@ func NewConnection(addr, user, pass string) *Connection {
 	}
 }
 
+func NewTLSConnection(addr, user, pass string, certPool *x509.CertPool) *Connection {
+	connection := NewConnection(addr, user, pass)
+	connection.dial = func(network, address string) (net.Conn, error) {
+		conn, err := net.DialTimeout(network, address, 5*time.Second)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO Save Info packet somewhere accessible
+		br := bufio.NewReaderSize(conn, 32768)
+		_, err = Parse(br)
+		if err != nil {
+			return conn, err
+		}
+
+		hostname, _, err := net.SplitHostPort(address)
+		if err != nil {
+			return conn, err
+		}
+
+		conn = tls.Client(conn, &tls.Config{
+			RootCAs: certPool,
+			ServerName: hostname,
+		})
+
+		tlsConn := conn.(*tls.Conn)
+		err = tlsConn.Handshake()
+		return tlsConn, err
+	}
+	return connection
+}
+
 type ConnectionInfo struct {
 	Addr     string
 	Username string
 	Password string
 	Dial     func(network, address string) (net.Conn, error)
+
+	CertPool *x509.CertPool
 }
 
 func (c *ConnectionInfo) ProvideConnection() (*Connection, error) {
-	conn := NewConnection(c.Addr, c.Username, c.Password)
+	var conn *Connection
+	if c.CertPool == nil || len(c.CertPool.Subjects()) == 0 {
```

Comment:
should we have the length check in this? since non-nil CertPool is non-default, it seems safer to say "if CertPool is present in any way... we should act securely". maybe